tf2onnx/optimizer/merge_duplicated_nodes_optimizer.py

```diff
@@ -0,0 +1,95 @@
+   for example, node a is input of node b and node c, and computation of node b, c are same such as "abs" op.
+   then b and c can be merged into one node to avoid duplicated computation
+"""
+
+from collections import defaultdict, namedtuple
+import logging
+
+from tf2onnx.optimizer.optimizer_base import GraphOptimizerBase
+
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger("tf2onnx.optimizer.merge_duplicated_nodes")
+
+
+# pylint: disable=logging-not-lazy,unused-argument,missing-docstring
+
+
+class MergeDuplicatedNodesOptimizer(GraphOptimizerBase):
+    """Remove duplicate nodes.
+    """
+    def __init__(self, debug=False):
+        super(MergeDuplicatedNodesOptimizer, self).__init__(debug)
+        self._name = "MergeDuplicatedNodesOptimizer"
+
+        self._graph_can_be_optimized = True
+        # define class used internally
+        self._key_to_group_nodes = namedtuple("key", "type input")
+
+    def _optimize(self, graph):
+        self._traversely_optimize_graph(graph, self._optimize_at_current_graph_level)
+        return graph
+
+    def _optimize_at_current_graph_level(self, graph):
+        while self._graph_can_be_optimized:
+            self._graph_can_be_optimized = False
+            self._merge_duplicated_nodes(graph)
+        return graph
+
+    def _merge_duplicated_nodes(self, graph):
+        # "duplicated" means: op_type, input and attribute are same
+        # while attr is un-hashable so doesn't include it when grouping nodes
+        nodes_groups = self._group_nodes_by_type_inputs(graph)
+        for key, nodes_group in nodes_groups.items():
+            if self._skip_node_type(nodes_group[0]):
+                continue
+            self._del_nodes_if_duplicated(nodes_group, graph)
+
+    def _group_nodes_by_type_inputs(self, graph):
+        res = defaultdict(list)
+        for node in graph.get_nodes():
+            res[self._key_to_group_nodes(node.type, tuple(node.input))].append(node)
+        return res
+
+    def _del_nodes_if_duplicated(self, nodes_group, graph):
+        # input and op type of nodes in same group are same,
+        # and if their attributes are also same then they are duplicated
+        while len(nodes_group) > 1:
+            unprocessed_node = []
+            nodes_to_process = [nodes_group[0]]
+            for node in nodes_group[1:]:
+                if node.attr == nodes_to_process[0].attr:
+                    nodes_to_process.append(node)
+                else:
+                    unprocessed_node.append(node)
+
+            self._merge_nodes_that_are_duplicated(nodes_to_process, graph)
+            nodes_group = unprocessed_node
+
+    def _merge_nodes_that_are_duplicated(self, nodes_to_process, graph):
+        # node's output may not all be used, so have to select the one that uses most of node's outputs
+        nodes_to_process.sort(key=self._len_of_node_output, reverse=True)
```

Comment:
i thought lengths will always be equal?

Reply:
if node's output is optional, then len is different